### ISSUE TITLE
1.1.2 breaks logging when using default credentials

### ISSUE THREAD

**Issue by imathews** [googleapis/nodejs-logging#53]
#### Environment details

  - OS: GKE, kubernetes 1.8.6
  - Node.js version: 8.6
  - @google-cloud/logging version: 1.1.2

#### Steps to reproduce

When upgrading from `1.1.1` to `1.1.2`, logging no longer works in our staging / prod environments — I don't see any of the logs in the Stackdriver interface, but otherwise it fails silently, with no corresponding errors, etc. I can confirm that reverting to `1.1.1`, without changing anything else, remedies the problem.

Interestingly, this all behaves fine on `1.1.2` in our dev environment (we still send logs to stackdriver for testing purposes). The only difference between dev and staging / prod is that we're using a keyfile in dev, while we rely on the default credentials within the nodes of our GKE cluster to authenticate in the other environments.

Let me know if I can provide any additional info to help troubleshoot, and many thanks in advance.

**Comment by stephenplusplus** [comment]
Thanks for reporting, and I'm very sorry for the delay.

@ofrobots I would guess that this problem relates to google-auto-auth and/or google-auth-library. In `@google-cloud/logging` 1.1.1, we used google-auto-auth@^0.7.1. In 1.1.2 (when the problem originates), we use google-auto-auth@^0.9.0. In that release span, g-a-a started using google-auth-library's "getCredentials" method. Is it possible that method doesn't support GKE environments?

**Comment by stephenplusplus** [comment]
Ping @JustinBeckwith :)

**Comment by stephenplusplus** [comment]
Hey @imathews, would you mind trying again? I believe we've caught the bug, and have released a fix under 1.1.3.

**Comment by imathews** [comment]
@stephenplusplus I just upgraded to 1.1.3; unfortunately, it looks like the problem is still there (don't see anything in logging, reverting to 1.1.1 brings it back) :/

Let me know if there's any other info I can provide to help with debugging.

**Comment by GuillaumeLachaud** [comment]
I think I am seeing something very similar when running on a GCE VM. Running locally works fine but the same sample code fails silently on GCE. I digged into it a little bit and it looks like the function `metadataAccessor` in gcp-metadata fails at triggering a callback in `Metadata.getGCEDescriptor` even though it get a successful response from the metadata service

**Comment by stephenplusplus** [comment]
@ofrobots @JustinBeckwith since you've been working a lot with GAL and gcp-metadata lately, can you think of any explanations?

**Comment by JustinBeckwith** [comment]
Aiight, I did a little digging here.  The `getCredentials` method is a weird beast.  I tested the method in a GKE cluster, and you get this back:

```js
{ client_email: '904342750300-compute@developer.gserviceaccount.com' }
```

@stephenplusplus are you expecting to get a private key back by chance?  I'm not 100% sure how you're using this method.  Why do you want `getCredentials` as opposed to just using adc?

**Comment by dan-j** [comment]
Just want to jump in on this issue because I'm having similar issues but different versions.

TL;DR; `1.1.2` works but `~1.1.3` doesn't.

I've been using `@google-cloud/logging@1.1.2` with no issues in test (keyfile defined in the environment variable `GOOGLE_APPLICATION_CREDENTIALS`) and production (using `admin.credentials.applicationDefault()`)

I recently noticed we weren't receiving logs in production and it turned out our Google Compute VM has been installing `1.1.4` instead of `1.1.2` and this results in no logs reaching StackDriver, and no errors either.

I've manually installed `1.1.2` and `1.1.3` in our production VM/container and verified `1.1.2` works as expected and neither `1.1.3` nor `1.1.4` work.

Any ideas how this could be breaking in `1.1.2` for some users and only working in `1.1.2` for others? I'm at home now but can provide the dependency tree tomorrow to see what transitive dependencies are being pulled in.

**Comment by stephenplusplus** [comment]
> but can provide the dependency tree tomorrow to see what transitive dependencies are being pulled in. - @dan-j

That would be great, thanks 👍 

> I'm not 100% sure how you're using this method. Why do you want getCredentials as opposed to just using adc? - @JustinBeckwith 

Actually, I think I falsely blamed that method. We aren't using it from Logging. Sorry to make you go through the work to debug in that direction :(

This comment seems more useful:

> I digged into it a little bit and it looks like the function `metadataAccessor` in gcp-metadata fails at triggering a callback in `Metadata.getGCEDescriptor` even though it get a successful response from the metadata service - @GuillaumeLachaud

Between Logging 1.1 -> 1.2, we went from gcp-metadata@^0.3 to ^0.4.

**Comment by stephenplusplus** [comment]
I've upgraded gcp-metadata and released a new google-auto-auth. Could everyone try re-installing/deploying and see if the new version fixes? (Logging@1.1.2 and up should get the new version of gcp-metadata)

**Comment by dan-j** [comment]
So my dependency trees using `1.1.2` and `1.1.4` is as follows:

I'm also wary of the fact I'm using Yarn in development which publishes my app as a private NPM package, and the production environment just installs the published package using `npm install <package>` (hence the tree-format difference below). I need to look into using a more deterministic solution...

`1.1.4` (not working):

```
+-- @google-cloud/logging-bunyan@0.7.0
| +-- @google-cloud/logging@1.1.4
| | +-- @google-cloud/common@0.16.1
| | | +-- array-uniq@1.0.3 deduped
| | | +-- arrify@1.0.1 deduped
| | | +-- concat-stream@1.6.0 deduped
| | | +-- create-error-class@3.0.2 deduped
| | | +-- duplexify@3.5.3 deduped
| | | +-- ent@2.2.0 deduped
| | | +-- extend@3.0.1 deduped
| | | +-- google-auto-auth@0.9.3 deduped
| | | +-- is@3.2.1 deduped
| | | +-- log-driver@1.2.5 deduped
| | | +-- methmeth@1.1.0 deduped
| | | +-- modelo@4.2.3 deduped
| | | +-- request@2.83.0 deduped
| | | +-- retry-request@3.3.1 deduped
| | | +-- split-array-stream@1.0.3 deduped
| | | +-- stream-events@1.0.2 deduped
| | | +-- string-format-obj@1.1.1 deduped
| | | `-- through2@2.0.3 deduped
| | +-- @google-cloud/common-grpc@0.5.5 deduped
| | +-- arrify@1.0.1 deduped
| | +-- eventid@0.1.2
| | | +-- d64@1.0.0
| | | `-- uuid@3.2.1 deduped
| | +-- extend@3.0.1 deduped
| | +-- gcp-metadata@0.5.0
| | | +-- axios@0.17.1
| | | | +-- follow-redirects@1.4.1
| | | | | `-- debug@3.1.0 deduped
| | | | `-- is-buffer@1.1.6
| | | +-- extend@3.0.1 deduped
| | | `-- retry-axios@0.3.0
| | +-- google-auto-auth@0.9.3
| | | +-- async@2.6.0 deduped
| | | +-- gcp-metadata@0.4.1
| | | | +-- extend@3.0.1 deduped
| | | | `-- retry-request@3.3.1 deduped
| | | +-- google-auth-library@0.12.0 deduped
| | | `-- request@2.83.0 deduped
| | +-- google-gax@0.14.5 deduped
| | +-- google-proto-files@0.15.1
| | | +-- globby@7.1.1 deduped
| | | +-- power-assert@1.4.4 deduped
| | | `-- protobufjs@6.8.5 deduped
| | +-- is@3.2.1 deduped
| | +-- lodash.merge@4.6.1
| | +-- protobufjs@6.8.5
| | | +-- @protobufjs/aspromise@1.1.2 deduped
| | | +-- @protobufjs/base64@1.1.2 deduped
| | | +-- @protobufjs/codegen@2.0.4 deduped
| | | +-- @protobufjs/eventemitter@1.1.0 deduped
| | | +-- @protobufjs/fetch@1.1.0 deduped
| | | +-- @protobufjs/float@1.0.2 deduped
| | | +-- @protobufjs/inquire@1.1.0 deduped
| | | +-- @protobufjs/path@1.1.2 deduped
| | | +-- @protobufjs/pool@1.1.0 deduped
| | | +-- @protobufjs/utf8@1.1.0 deduped
| | | +-- @types/long@3.0.32 deduped
| | | +-- @types/node@8.9.4 deduped
| | | `-- long@4.0.0
| | +-- pumpify@1.4.0
| | | +-- duplexify@3.5.3 deduped
| | | +-- inherits@2.0.3 deduped
| | | `-- pump@2.0.1
| | |   +-- end-of-stream@1.4.1 deduped
| | |   `-- once@1.4.0 deduped
| | +-- snakecase-keys@1.1.1
| | | +-- map-obj@2.0.0
| | | `-- to-snake-case@0.1.2
| | |   `-- to-space-case@0.1.2
| | |     `-- to-no-case@0.1.1
| | +-- stream-events@1.0.2 deduped
| | +-- string-format-obj@1.1.1 deduped
| | `-- through2@2.0.3 deduped
| `-- extend@3.0.1 deduped
```

`1.1.2` (is working):

```
├─ @google-cloud/logging-bunyan@0.7.0
│  ├─ @google-cloud/logging@^1.0.0
│  └─ extend@^3.0.1
├─ @google-cloud/logging@1.1.2
│  ├─ @google-cloud/common-grpc@^0.5.0
│  ├─ @google-cloud/common@^0.15.0
│  ├─ arrify@^1.0.0
│  ├─ eventid@^0.1.0
│  ├─ extend@^3.0.1
│  ├─ gcp-metadata@^0.4.0
│  ├─ gcp-metadata@0.4.1
│  │  ├─ extend@^3.0.0
│  │  └─ retry-request@^3.1.0
│  ├─ google-auto-auth@^0.9.0
│  ├─ google-auto-auth@0.9.1
│  │  ├─ async@^2.3.0
│  │  ├─ gcp-metadata@^0.4.1
│  │  ├─ google-auth-library@^0.12.0
│  │  └─ request@^2.79.0
│  ├─ google-gax@^0.14.2
│  ├─ google-gax@0.14.3
│  │  ├─ extend@^3.0.0
│  │  ├─ gcp-metadata@0.3.1
│  │  │  ├─ extend@^3.0.0
│  │  │  └─ retry-request@^3.0.0
│  │  ├─ globby@^6.1.0
│  │  ├─ google-auto-auth@^0.8.0
│  │  ├─ google-auto-auth@0.8.2
│  │  │  ├─ async@^2.3.0
│  │  │  ├─ gcp-metadata@^0.3.0
│  │  │  ├─ google-auth-library@^0.12.0
│  │  │  └─ request@^2.79.0
│  │  ├─ google-proto-files@^0.13.1
│  │  ├─ google-proto-files@0.13.1
│  │  ├─ grpc@~1.7.2
│  │  ├─ is-stream-ended@^0.1.0
│  │  ├─ lodash@^4.17.2
│  │  ├─ process-nextick-args@^1.0.7
│  │  ├─ protobufjs@^6.8.0
│  │  ├─ readable-stream@^2.2.2
│  │  └─ through2@^2.0.3
│  ├─ google-proto-files@^0.14.0
│  ├─ is@^3.0.1
│  ├─ lodash.merge@^4.6.0
│  ├─ protobufjs@^6.8.0
│  ├─ pumpify@^1.3.5
│  ├─ snakecase-keys@^1.1.0
│  ├─ stream-events@^1.0.1
│  ├─ string-format-obj@^1.0.0
│  └─ through2@^2.0.3
```

**Comment by dan-j** [comment]
On top of my previous comment, that dependency tree for 1.1.2 may not be representative of what I was using last night when I got 1.1.2 to work...

That dependency tree was created this morning and looks like the production environment redeployed since your release and different transitive dependency versions have been pulled in. Now 1.1.2 isn't working for me, but not 100% if it's my error or a dependencies. I'm still looking into it...

**Comment by dan-j** [comment]
Ok, `1.1.2` works for me using Yarn. It was my error yesterday where I couldn't get `1.1.2` to work. I was using the `@google-cloud/logging-bunyan@0.7.0` which was installing `@google-cloud/logging@1.1.4` in a nested node_modules directory, so even though I had `1.1.2` installed as a direct dependency logging-bunyan was still using `1.1.4`.

For people using Yarn, I've used the following `resolutions` block in my package.json:

```
  "resolutions": {
    "@google-cloud/logging": "1.1.2"
  },
```

Here's the output of `yarn list` using `1.1.2`, `1.1.3` and `1.1.4`.

[yarn-list-1.1.4.txt](https://github.com/googleapis/nodejs-logging/files/1749957/yarn-list-1.1.4.txt)
[yarn-list-1.1.3.txt](https://github.com/googleapis/nodejs-logging/files/1749951/yarn-list-1.1.3.txt)
[yarn-list-1.1.2.txt](https://github.com/googleapis/nodejs-logging/files/1749952/yarn-list-1.1.2.txt)

**Comment by imathews** [comment]
@stephenplusplus I tried updating on Friday to `1.1.4`; unfortunately, this behavior persisted.

**Comment by Cory-Simon** [comment]
Environment details

OS: Container-Optimized OS v64
Image: gcr.io/google_appengine/nodejs
Node.js version: 8.94
 version: 1.1.1 - 1.1.4

Not sure how much this helps, but we are having the same issue using @google-cloud/logging-winston.

Ran the code from the README using each version of @google-gloud/logging for 1.1.1 to 1.1.4 in a clean environment described above and only saw messages successfully for 1.1.1 & 1.1.2

Here is the output for npm list for each version:

[1_1_1.txt](https://github.com/googleapis/nodejs-logging/files/1768358/1_1_1.txt)
[1_1_2.txt](https://github.com/googleapis/nodejs-logging/files/1768359/1_1_2.txt)
[1_1_3.txt](https://github.com/googleapis/nodejs-logging/files/1768360/1_1_3.txt)
[1_1_4.txt](https://github.com/googleapis/nodejs-logging/files/1768361/1_1_4.txt)

**Comment by Cory-Simon** [comment]
Pulling on the thread from the comment by @GuillaumeLachaud the [gcp-metadata](https://github.com/stephenplusplus/gcp-metadata/) library did update from `v0.3.1` to `v0.5.0` when google-cloud/logging updated from `v1.1.2` to `v1.1.3`. Running node while in inspect mode (in a container on GCE) the behavior that @GuillaumeLachaud describes is what I observe as well.

Looking at a diff of the 2 versions the implementation did change significantly. I'm not sure if those changes are the cause of the bug here however. My JS skills aren't quite at that level. Reporting my findings so that someone with more skills might make that determination.

[Diff of the 2 versions](https://github.com/stephenplusplus/gcp-metadata/compare/d8d9878...1841f3c)

**Comment by laanuk** [comment]
Also using @google-cloud/logging-winston - but I can't get it to work on GKE on any version from 1.1.1 to 1.1.4. They show up in the gcp logs viewer if i run the code locally.

**Comment by stephenplusplus** [comment]
I believe the fix has just been released under 1.1.5. Please try it out and let me know how it goes!

**Comment by dan-j** [comment]
I've attempted 1.1.6 but then I get this error, downgrading to 1.1.2 again for now...

```
(node:40) UnhandledPromiseRejectionWarning: RangeError: Maximum call stack size exceeded
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:550:31)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
    at replaceProjectIdToken (/home/node/app/node_modules/@google-cloud/common/src/util.js:560:22)
(node:40) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
(node:40) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

```
├─ @google-cloud/logging@1.1.6
│  ├─ @google-cloud/common-grpc@^0.6.0
│  ├─ @google-cloud/common-grpc@0.6.0
│  │  ├─ @google-cloud/common@^0.16.1
│  │  ├─ dot-prop@^4.2.0
│  │  ├─ duplexify@^3.5.1
│  │  ├─ extend@^3.0.1
│  │  ├─ grpc@~1.9.1
│  │  ├─ is@^3.2.0
│  │  ├─ modelo@^4.2.0
│  │  ├─ retry-request@^3.3.1
│  │  └─ through2@^2.0.3
│  ├─ @google-cloud/common@^0.16.1
│  ├─ arrify@^1.0.0
│  ├─ camelcase@2.1.1
│  ├─ cliui@3.2.0
│  │  ├─ string-width@^1.0.1
│  │  ├─ strip-ansi@^3.0.1
│  │  └─ wrap-ansi@^2.0.0
│  ├─ eventid@^0.1.0
│  ├─ extend@^3.0.1
│  ├─ gcp-metadata@^0.6.1
│  ├─ globby@8.0.1
│  │  ├─ array-union@^1.0.1
│  │  ├─ dir-glob@^2.0.0
│  │  ├─ fast-glob@^2.0.2
│  │  ├─ glob@^7.1.2
│  │  ├─ ignore@^3.3.5
│  │  ├─ pify@^3.0.0
│  │  └─ slash@^1.0.0
│  ├─ google-auto-auth@^0.9.0
│  ├─ google-gax@^0.15.0
│  ├─ google-gax@0.15.0
│  │  ├─ extend@^3.0.0
│  │  ├─ globby@^8.0.0
│  │  ├─ google-auto-auth@^0.9.0
│  │  ├─ google-proto-files@^0.15.0
│  │  ├─ grpc@~1.9.1
│  │  ├─ is-stream-ended@^0.1.0
│  │  ├─ lodash@^4.17.2
│  │  ├─ protobufjs@^6.8.0
│  │  ├─ readable-stream@^2.2.2
│  │  └─ through2@^2.0.3
│  ├─ google-proto-files@^0.15.0
│  ├─ grpc@1.9.1
│  │  ├─ lodash@^4.15.0
│  │  ├─ nan@^2.0.0
│  │  ├─ node-pre-gyp@^0.6.39
│  │  ├─ protobufjs@^5.0.0
│  │  └─ protobufjs@5.0.2
│  │     ├─ ascli@~1
│  │     ├─ bytebuffer@~5
│  │     ├─ glob@^7.0.5
│  │     └─ yargs@^3.10.0
│  ├─ is@^3.0.1
│  ├─ lodash.merge@^4.6.0
│  ├─ os-locale@1.4.0
│  │  └─ lcid@^1.0.0
│  ├─ protobufjs@^6.8.0
│  ├─ pumpify@^1.3.5
│  ├─ snakecase-keys@^1.1.0
│  ├─ stream-events@^1.0.1
│  ├─ string-format-obj@^1.0.0
│  ├─ through2@^2.0.3
│  └─ yargs@3.32.0
│     ├─ camelcase@^2.0.1
│     ├─ cliui@^3.0.3
│     ├─ decamelize@^1.1.1
│     ├─ os-locale@^1.4.0
│     ├─ string-width@^1.0.1
│     ├─ window-size@^0.1.4
│     └─ y18n@^3.2.0
```

**Comment by stephenplusplus** [comment]
Thanks, I've moved that over here and will take a look now: https://github.com/googleapis/nodejs-logging/issues/86
